src/attr/_make.py

```diff
@@ -900,12 +911,106 @@ def wrap(cls):
         return wrap(maybe_cls)
 
 
+singletons = defaultdict(dict)
+initialized = set()
+
+
+def _s_get(arg_len, args, kwargs, attr, idx):
+    if attr.name in kwargs:
+        return kwargs[attr.name]
+    elif idx < arg_len:
+        return args[idx]
+    elif attr.default != NOTHING:
+        return attr.default
+    else:
+        raise ValueError("attribute not initialized")
+
+
+def _empty(*args, **kwargs):
+    pass
+
+
+def singleton(maybe_cls=None):
+    """
+    Decorator which transforms an attrs class or namedtuple into a singleton.
+    This means that any time you instantiate that class with the same
+    arguments, you are garaunteed to get the same object.
+    """
+
+    def wrap(_cls):
+        if _has_frozen_superclass(_cls):
+
+            @staticmethod
+            def __new__(cls, *arg, **kwarg):
+                """
+                attrs-generated __new__ for a singleton, frozen attrs class
+                """
+                arglen = len(arg)
+                iter = enumerate(fields(cls))
+                key = tuple(
+                    _s_get(arglen, arg, kwarg, a, i) for i, a in iter if a.init
+                )
+                class_listings = singletons[cls]
+                if key not in class_listings:
+                    class_listings[key] = super(cls, cls).__new__(cls)
+                return class_listings[key]
+
+        elif issubclass(_cls, tuple):
+
+            @staticmethod
+            def __new__(cls, *args, **kwargs):
+                """
+                attrs-generated __new__ for a namedtuple-based singleton class
+                """
+                mapping = {}
+                mapping.update(getattr(cls, "_field_defaults", {}))
+                mapping.update(kwargs)
+                for key, val in zip(cls._fields, args):
+                    mapping[key] = val
+                t_key = tuple(mapping[key] for key in cls._fields)
+                class_listings = singletons[cls]
+                if t_key not in class_listings:
+                    class_listings[t_key] = super(cls, cls).__new__(cls, t_key)
+                return class_listings[t_key]
+
+        else:
+            raise TypeError("Must be run on frozen attrs class or namedtuple")
+
+        def __init__(self, *args, **kwargs):
+            try:
+                if self in initialized:
+                    return
+                prev_init(self, *args, **kwargs)
```

Comment:
I don't like that these lines are repeated, but I could not easily think of a better mechanism for this. I am open to suggestions on improving this mechanism.

Reply:
Maybe it would be good to have:

```python
def __init__(self, *args, **kwargs):
    try:
        if self not in initialized:
            raise AttributeError()
    except AttributeError:
        prev_init(self, *args, **kwargs)
        initialized.add(self)
```

Reply:
That is slightly less efficient, but it's a one-time cost per object, and only really affects pypy (I think)